### PR TITLE
WIP: Safer tests

### PR DESCRIPTION
**Pull request by rmg**
I started cleaning up the tests in order to debug the windows failures, but haven't had time to finish yet.

@superkhau and @jannyHou have each contacted me separately about the failures, so I'm opening this PR to get more eyes.

### CODE REVIEW

**Comment by jannyHou** [comment]
hey @rmg @bajtos I test on one a linux vm and manage to reproduce 4 errors.
(Configue: OS: ub14x86, node 4.1.0, npm 2.14.3)
I fixed them by **removing `strong-remoting`** under dependant `loopback-connector-remote`'s folder.

When I clone loopback from github and npm install it, **loopback-connector-remote** installs **strong-remoting 2.24.5** automatically under its node_modules folder which causes the failure. 
I guess it calls that one instead of the strong-remoting in root node_modules folder.
There is also a `loopback-datasource-juggler 2.24.22` there but it doesn't affect the results.

I am not sure are they necessarily installed as nested modules of loopback-connector-remote:) but from the version number they look old and npm test could pass after I remove them.


Error details:
```
1) RemoteConnector should support the save method:

      Uncaught AssertionError: false == true
      + expected - actual

      -false
      +true

      at HttpInvocation.<anonymous> (test/remote-connector.test.js:68:7)
      at node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:131:18
      at execStack (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:460:7)
      at RemoteObjects.execHooks (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:464:10)
      at HttpInvocation.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:129:15)
      at HttpInvocation.transformResponse (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:337:12)
      at Request._callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:261:10)
      at Request.self.callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:198:22)
      at Request.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:1035:10)
      at IncomingMessage.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:962:12)

  2) RemoteConnector Model Tests Model.create([data], [callback]) Create an instance of Model with given data and save to the attached data source:

      Uncaught AssertionError: false == true
      + expected - actual

      -false
      +true

      at HttpInvocation.<anonymous> (test/util/model-tests.js:136:11)
      at node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:131:18
      at execStack (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:460:7)
      at RemoteObjects.execHooks (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:464:10)
      at HttpInvocation.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:129:15)
      at HttpInvocation.transformResponse (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:337:12)
      at Request._callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:261:10)
      at Request.self.callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:198:22)
      at Request.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:1035:10)
      at IncomingMessage.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:962:12)

  3) RemoteConnector Model Tests model.updateAttributes(data, [callback]) Save specified attributes to the attached data source:
     Uncaught TypeError: user.updateAttributes is not a function
      at HttpInvocation.<anonymous> (test/util/model-tests.js:160:16)
      at node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:131:18
      at execStack (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:460:7)
      at RemoteObjects.execHooks (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:464:10)
      at HttpInvocation.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:129:15)
      at HttpInvocation.transformResponse (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:337:12)
      at Request._callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:261:10)
      at Request.self.callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:198:22)
      at Request.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:1035:10)
      at IncomingMessage.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:962:12)

  4) Replication over REST conflict resolution with model-level permissions as user with REPLICATE-only permissions allows reverse resolve() on the client:
     Uncaught TypeError: target.toObject is not a function
      at common/models/change.js:706:16
      at done (common/models/change.js:604:7)
      at node_modules/async/lib/async.js:251:17
      at done (node_modules/async/lib/async.js:132:19)
      at node_modules/async/lib/async.js:32:16
      at node_modules/async/lib/async.js:248:21
      at node_modules/async/lib/async.js:572:34
      at HttpInvocation.<anonymous> (common/models/change.js:598:9)
      at node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:131:18
      at execStack (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:460:7)
      at RemoteObjects.execHooks (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/remote-objects.js:464:10)
      at HttpInvocation.<anonymous> (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/rest-adapter.js:129:15)
      at HttpInvocation.transformResponse (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:337:12)
      at Request._callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/lib/http-invocation.js:261:10)
      at Request.self.callback (node_modules/loopback-connector-remote/node_modules/strong-remoting/node_modules/request/request.js:198:22)

```

**Comment by bajtos** [comment]
@jannyHou thank you for the investigation and a detailed report.

loopback-connector-remote depends on strong-remoting (and loopback-datasource-juggler) because it uses some of the code/functionality provided by these modules.

We were aware that juggler is designed as a singleton (i.e. there should be only one instance of the module in the dependency tree), there is a task to fix it - https://github.com/strongloop/loopback-connector-remote/issues/5

I am quite surprised that strong-remoting suffers from the same problem too. We need to investigate this more to better understand why two instances of strong-remoting don't work well.

Here is the reason why this problem is not observed on the 2.x branch: because both `loopback` and `loopback-connector-remote` depend on the latest 2.x version of `strong-remoting`, `npm install` will skip installation of `loopback-connector-remote/node_modules/strong-remoting` and let `loopback-connector-remote` use the copy of `strong-remoting` installed directly in loopback's `node_modules`.

**Comment by bajtos** [comment]
I identified the problem. strong-remoting has a global registry of `Dynamic` converters that are used to convert responses e.g. from a plain object to a model instance. LoopBack registers all models with it's instance of strong-remoting. However, the remoting connector has its own `Dynamic` registry that is not populated with application models.

@jannyHou fixing this problem is non-trivial, I'll work on this myself - see https://github.com/strongloop/loopback-connector-remote/issues/34

**Comment by bajtos** [comment]
@slnode test please

**Comment by bajtos** [comment]
strongloop/loopback-connector-remote#34 has been landed, let's check how CI likes that

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# LoopBack
+# LoopBack1
```
Comment (bajtos) [inline README.md]: Looks like a type/forgotten change, please remove.

File: test/remoting.integration.js [hunk 2]
```diff
@@ -6,6 +6,12 @@ var app = require(path.join(SIMPLE_APP, 'server/server.js'));
 var assert = require('assert');
 
 describe('remoting - integration', function() {
+  before(function(done) {
+    if (app.booting) {
+      return app.once('booted', done);
+    }
+    done();
+  });
```
Comment (bajtos) [inline test/remoting.integration.js]: Can we move this block into `lt.beforeEach.withApp`, so that we don't have to duplicate it?

Comment (bajtos) [inline test/remoting.integration.js]: I would simply move it into the `beforeEach` hook code here: https://github.com/strongloop/loopback/blob/0ad150cb6ee1501d5bcbdb077ee5621cdb64c407/test/helpers/loopback-testing-helper.js#L21-L28

**Comment by jannyHou** [comment]
@slnode test please

**Comment by jannyHou** [comment]
strong-remoting@2.24.5 node_modules\strong-remoting
├── eventemitter2@0.4.14
├── js2xmlparser@0.1.9
├── qs@2.4.2
├── traverse@0.6.6
├── cors@2.7.1 (vary@1.1.0)
├── sse@0.0.6 (options@0.0.6)
├── xml2js@0.4.16 (sax@1.1.5, xmlbuilder@4.2.1)
├── jayson@1.2.2 (eyes@0.1.8, commander@1.3.2, lodash@3.6.0, JSONStream@1.0.3)
├── mux-demux@3.7.9 (json-buffer@2.0.11, duplex@1.0.0, stream-serializer@1.1.2,                                       xtend@1.0.3, through@2.3.8, stream-combiner@0.0.2, msgpack-stream@0.0.12)
└── request@2.69.0 (forever-agent@0.6.1, tunnel-agent@0.4.2, is-typedarray@1.0.0                                      , oauth-sign@0.8.0, aws-sign2@0.6.0, caseless@0.11.0, stringstream@0.0.5, isstre                                      am@0.1.2, json-stringify-safe@5.0.1, extend@3.0.0, tough-cookie@2.2.1, node-uuid                                      @1.4.7, qs@6.0.2, combined-stream@1.0.5, mime-types@2.1.9, aws4@1.2.1, form-data                                      @1.0.0-rc3, hawk@3.1.3, bl@1.0.1, http-signature@1.1.1, har-validator@2.0.6)

Strong-remoting 2.24.5 makes all tests pass on my local.
OS: windows 64 bit
node: 4.2.6
npm: 2.14.12

**Comment by jannyHou** [comment]
@slnode test please

**Comment by jannyHou** [comment]
Errors come from [`test/hidden-properties.test.js`](https://github.com/strongloop/loopback/blob/master/test/hidden-properties.test.js#L37), I commented `.expect('Content-Type', /json/)` to expose the real error:
```
expected 200 "OK", got 404 "Not Found"
```
So the pr for "content type" is fine, the error comes from endpoint itself
The single test passes when I run `mocha -g 'should hide a property remotely'`
I am also confused why the test failures are duplicate(6 errors in total but actually only 3 errors, other 3 are duplicate).


Detailed error information with `longjohn`:
```
1)  "before each" hook:
     TypeError: Cannot read property 'currentRetry' of undefined
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:536:27
      at Runner.runTest (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mo                                      dules\mocha\lib\runner.js:424:5)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:533:12
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:342:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:352:7
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:284:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:315:7
      at done (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runnable.js:276:5)
      at callFn (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mo                                      cha\lib\runnable.js:333:7)
      at Hook.Runnable.run (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node                                      _modules\mocha\lib\runnable.js:308:7)
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:298:10)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)
  ---------------------------------------------
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:293:12)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)

  2) hidden properties should hide a property remotely:
     Error: expected 200 "OK", got 404 "Not Found"
      at Test.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\supertest\lib\test.js:205:15)
      at Server.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mod                                      ules\supertest\lib\test.js:132:12)
      at emitCloseNT (net.js:1524:8)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Server.close (net.js:1480:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\supertest\                                      lib\test.js:127:31
      at Test.Request.callback (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\                                      node_modules\supertest\node_modules\superagent\lib\node\index.js:746:30)
      at Test.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_                                      modules\supertest\node_modules\superagent\lib\node\index.js:135:10)
      at IncomingMessage.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loo                                      pback\node_modules\supertest\node_modules\superagent\lib\node\index.js:938:12)
      at endReadableNT (_stream_readable.js:905:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Test.Request (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modu                                      les\supertest\node_modules\superagent\lib\node\index.js:134:8)
      at new Test (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\                                      supertest\lib\test.js:29:11)
      at Object.obj.(anonymous function) [as get] (C:\Users\IBM_ADMIN\Desktop\ci                                      -loopback\loopback\node_modules\supertest\index.js:25:14)
      at Context.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\te                                      st\hidden-properties.test.js:39:9)
      at callFnAsync (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\mocha\lib\runnable.js:338:8)
      at Test.Runnable.run (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node                                      _modules\mocha\lib\runnable.js:290:7)
      at Runner.runTest (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mo                                      dules\mocha\lib\runner.js:422:10)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:533:12
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:342:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:352:7
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:284:14)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)

  3) hidden properties should hide a property remotely:
     Error: expected 200 "OK", got 404 "Not Found"
      at Test.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\supertest\lib\test.js:205:15)
      at Server.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mod                                      ules\supertest\lib\test.js:132:12)
      at emitCloseNT (net.js:1524:8)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Server.close (net.js:1480:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\supertest\                                      lib\test.js:127:31
      at Test.Request.callback (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\                                      node_modules\supertest\node_modules\superagent\lib\node\index.js:746:30)
      at Test.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_                                      modules\supertest\node_modules\superagent\lib\node\index.js:135:10)
      at IncomingMessage.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loo                                      pback\node_modules\supertest\node_modules\superagent\lib\node\index.js:938:12)
      at endReadableNT (_stream_readable.js:905:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Test.Request (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modu                                      les\supertest\node_modules\superagent\lib\node\index.js:134:8)
      at new Test (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\                                      supertest\lib\test.js:29:11)
      at Object.obj.(anonymous function) [as get] (C:\Users\IBM_ADMIN\Desktop\ci                                      -loopback\loopback\node_modules\supertest\index.js:25:14)
      at Context.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\te                                      st\hidden-properties.test.js:39:9)
      at callFnAsync (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\mocha\lib\runnable.js:338:8)
      at Test.Runnable.run (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node                                      _modules\mocha\lib\runnable.js:290:7)
      at Runner.runTest (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mo                                      dules\mocha\lib\runner.js:422:10)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:533:12
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:342:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:352:7
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:284:14)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)

  4) hidden properties should hide a property of nested models:
     Error: expected 200 "OK", got 404 "Not Found"
      at Test.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\supertest\lib\test.js:205:15)
      at Server.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mod                                      ules\supertest\lib\test.js:132:12)
      at emitCloseNT (net.js:1524:8)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Server.close (net.js:1480:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\supertest\                                      lib\test.js:127:31
      at Test.Request.callback (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\                                      node_modules\supertest\node_modules\superagent\lib\node\index.js:746:30)
      at Test.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_                                      modules\supertest\node_modules\superagent\lib\node\index.js:135:10)
      at IncomingMessage.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loo                                      pback\node_modules\supertest\node_modules\superagent\lib\node\index.js:938:12)
      at endReadableNT (_stream_readable.js:905:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Test.Request (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modu                                      les\supertest\node_modules\superagent\lib\node\index.js:134:8)
      at new Test (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\                                      supertest\lib\test.js:29:11)
      at Object.obj.(anonymous function) [as get] (C:\Users\IBM_ADMIN\Desktop\ci                                      -loopback\loopback\node_modules\supertest\index.js:25:14)
      at Context.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\te                                      st\hidden-properties.test.js:53:8)
      at callFnAsync (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\mocha\lib\runnable.js:338:8)
      at Test.Runnable.run (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node                                      _modules\mocha\lib\runnable.js:290:7)
      at Runner.runTest (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mo                                      dules\mocha\lib\runner.js:422:10)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:533:12
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:342:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:352:7
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:284:14)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)

  5) hidden properties should hide a property of nested models:
     Error: expected 200 "OK", got 404 "Not Found"
      at Test.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\supertest\lib\test.js:205:15)
      at Server.assert (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mod                                      ules\supertest\lib\test.js:132:12)
      at emitCloseNT (net.js:1524:8)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Server.close (net.js:1480:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\supertest\                                      lib\test.js:127:31
      at Test.Request.callback (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\                                      node_modules\supertest\node_modules\superagent\lib\node\index.js:746:30)
      at Test.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_                                      modules\supertest\node_modules\superagent\lib\node\index.js:135:10)
      at IncomingMessage.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loo                                      pback\node_modules\supertest\node_modules\superagent\lib\node\index.js:938:12)
      at endReadableNT (_stream_readable.js:905:12)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\continuati                                      on-local-storage\node_modules\async-listener\glue.js:188:31
  ---------------------------------------------
      at Test.Request (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modu                                      les\supertest\node_modules\superagent\lib\node\index.js:134:8)
      at new Test (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\                                      supertest\lib\test.js:29:11)
      at Object.obj.(anonymous function) [as get] (C:\Users\IBM_ADMIN\Desktop\ci                                      -loopback\loopback\node_modules\supertest\index.js:25:14)
      at Context.<anonymous> (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\te                                      st\hidden-properties.test.js:53:8)
      at callFnAsync (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modul                                      es\mocha\lib\runnable.js:338:8)
      at Test.Runnable.run (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node                                      _modules\mocha\lib\runnable.js:290:7)
      at Runner.runTest (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mo                                      dules\mocha\lib\runner.js:422:10)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:533:12
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:342:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:352:7
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:284:14)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)

  6)  "before each" hook:
     TypeError: Cannot read property 'currentRetry' of undefined
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:536:27
      at Runner.runTest (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_mo                                      dules\mocha\lib\runner.js:424:5)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:533:12
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:342:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:352:7
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:284:14)
      at C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mocha\lib\                                      runner.js:315:7
      at done (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runnable.js:276:5)
      at callFn (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\mo                                      cha\lib\runnable.js:333:7)
      at Hook.Runnable.run (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node                                      _modules\mocha\lib\runnable.js:308:7)
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:298:10)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)
  ---------------------------------------------
      at next (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback\node_modules\moch                                      a\lib\runner.js:293:12)
      at Immediate._onImmediate (C:\Users\IBM_ADMIN\Desktop\ci-loopback\loopback                                      \node_modules\mocha\lib\runner.js:320:5)

```

**Comment by richardpringle** [comment]
To me, it looks like there is an issue with **supertest**. I have seen this issue pop up before although I can't remember exactly how it gets resolved.

I'm guessing there is an issue booting up the server somewhere. In the past I've seen that for whatever reason, supertest does not "stop" the app instance properly, which would explain why the test is attempted multiple time, and why it passes on it's own but not with the other tests.

I'm not familiar with supertest but if @rmg or @bajtos are, maybe you could explain this?

In the meantime, I will look into supertest and try to figure out what's going on.

**Comment by jannyHou** [comment]
@richardpringle thanks! It makes sense why there are duplicate failures. Yesterday I searched 'Cannot read property 'currentRetry' of undefined' and got few result, while now it's pretty clear. I will look into **supertest** and see how to solve the conflict.

**Comment by bajtos** [comment]
Related: https://github.com/strongloop/loopback/pull/2034

**Comment by bajtos** [comment]
I believe all problems should be fixed by now, see #2036, #/2035, #2034, and #2030. I am closing this PR as no longer relevant.

**Comment by jannyHou** [comment]
@bajtos thank you!
